**You may need to install [OpenCV](https://pypi.python.org/pypi/opencv-python) and [scikit-video](http://www.scikit-video.org/stable/).**

In [13]:
import keras
import numpy as np
import io
import base64
from IPython.display import HTML
import skvideo.io
import cv2
import json

from keras.models import Sequential,model_from_json
from keras.layers.core import Dense
from keras.optimizers import sgd
from keras.layers import Conv2D, MaxPooling2D, Flatten, Activation, AveragePooling2D,Reshape,BatchNormalization

# MiniProject on Deep Reinforcement Learning

__Notations__: $E_p$ is the expectation under probability $p$. Please justify each of your answer and widely comment your code.

# Context

In a reinforcement learning algorithm, we modelize each step $t$ as an action $a_t$ obtained from a state $s_t$, i.e. $\{(a_{t},s_{t})_{t\leq T}\}$ having the Markov property. We consider a discount factor $\gamma \in [0,1]$ that ensures convergence. The goal is to find among all the policies $\pi$, one that maximizes the expected reward:

\begin{equation*}
R(\pi)=\sum_{t\leq T}E_{p^{\pi}}[\gamma^t r(s_{t},a_{t})] \> ,
\end{equation*}

where: 
\begin{equation*}p^{\pi}(a_{0},a_{1},s_{1},...,a_{T},s_{T})=p(a_{0})\prod_{t=1}^{T}\pi(a_{t}|s_{t})p(s_{t+1}|s_{t},a_{t}) \> .
\end{equation*}

We note the $Q$-function:

\begin{equation*}Q^\pi(s,a)=E_{p^{\pi}}[\sum_{t\leq T}\gamma^{t}r(s_{t},a_{t})|s_{0}=s,a_{0}=a] \> .
\end{equation*}

Thus, the optimal Q function is:
\begin{equation*}
Q^*(s,a)=\max_{\pi}Q^\pi(s,a) \> .
\end{equation*}

In this project, we will apply the deep reinforcement learning techniques to a simple game: an agent will have to learn from scratch a policy that will permit it maximizing a reward.

## The environment, the agent and the game

### The environment

```Environment``` is an abstract class that represents the states, rewards, and actions to obtain the new state.

In [2]:
class Environment(object):
    def __init__(self):
        pass

    def act(self, act):
        """
        One can act on the environment and obtain its reaction:
        - the new state
        - the reward of the new state
        - should we continue the game?

        :return: state, reward, game_over
        """
        pass


    def reset(self):
        """
        Reinitialize the environment to a random state and returns
        the original state

        :return: state
        """
        pass
    
    def draw(self):
        """
        Visualize in the console or graphically the current state
        """
        pass

The method ```act``` allows to act on the environment at a given state $s_t$ (stored internally), via action $a_t$. The method will return the new state $s_{t+1}$, the reward $r(s_{t},a_{t})$ and determines if $t\leq T$ (*game_over*).

The method ```reset``` simply reinitializes the environment to a random state $s_0$.

The method ```draw``` displays the current state $s_t$ (this is useful to check the behavior of the Agent).

We modelize $s_t$ as a tensor, while $a_t$ is an integer.

### The Agent

The goal of the ```Agent``` is to interact with the ```Environment``` by proposing actions $a_t$ obtained from a given state $s_t$ to attempt to maximize its __reward__ $r(s_t,a_t)$. We propose the following abstract class:

In [3]:
class Agent(object):
    def __init__(self, epsilon=0.1, n_action=4):
        self.epsilon = epsilon
        self.n_action = n_action
    
    def set_epsilon(self,e):
        self.epsilon = e

    def act(self,s,train=True):
        """ This function should return the next action to do:
        an integer between 0 and 4 (not included) with a random exploration of epsilon"""
        if train:
            if np.random.rand() <= self.epsilon:
                a = np.random.randint(0, self.n_action, size=1)[0]
            else:
                a = self.learned_act(s)
        else: # in some cases, this can improve the performance.. remove it if poor performances
            a = self.learned_act(s)

        return a

    def learned_act(self,s):
        """ Act via the policy of the agent, from a given state s
        it proposes an action a"""
        pass

    def reinforce(self, s, n_s, a, r, game_over_):
        """ This function is the core of the learning algorithm. 
        It takes as an input the current state s_, the next state n_s_
        the action a_ used to move from s_ to n_s_ and the reward r_.
        
        Its goal is to learn a policy.
        """
        pass

    def save(self):
        """ This function returns basic stats if applicable: the
        loss and/or the model"""
        pass

    def load(self):
        """ This function allows to restore a model"""
        pass

***
__Question 1__:
Explain the function act. Why is ```epsilon``` essential?

The function act gives the next action that will be performed by the Agent given the state s.
The epsilon parameter allow the control of the exploration of strategies by the Agent. Indeed, this is the $\epsilon$-greedy strategy : 
- a value of 0 gives the best strategie with the best reward untill the instant t. 
- a nonnegative value allows new strategies to be explored

***
### The Game

The ```Agent``` and the ```Environment``` work in an interlaced way as in the following (take some time to understand this code as it is the core of the project)

```python

epoch = 300
env = Environment()
agent = Agent()


# Number of won games
score = 0
loss = 0


for e in range(epoch):
    # At each epoch, we restart to a fresh game and get the initial state
    state = env.reset()
    # This assumes that the games will end
    game_over = False

    win = 0
    lose = 0
    
    while not game_over:
        # The agent performs an action
        action = agent.act(state)

        # Apply an action to the environment, get the next state, the reward
        # and if the games end
        prev_state = state
        state, reward, game_over = env.act(action)

        # Update the counters
        if reward > 0:
            win = win + reward
        if reward < 0:
            lose = lose -reward

        # Apply the reinforcement strategy
        loss = agent.reinforce(prev_state, state,  action, reward, game_over)

    # Save as a mp4
    if e % 10 == 0:
        env.draw(e)

    # Update stats
    score += win-lose

    print("Epoch {:03d}/{:03d} | Loss {:.4f} | Win/lose count {}/{} ({})"
          .format(e, epoch, loss, win, lose, win-lose))
    agent.save()
```

# The game, *eat cheese*

A rat runs on an island and tries to eat as much as possible. The island is subdivided into $N\times N$ cells, in which there are cheese (+0.5) and poisonous cells (-1). The rat has a visibility of 2 cells (thus it can see $5^2$ cells). The rat is given a time $T$ to accumulate as much food as possible. It can perform 4 actions: going up, down, left, right. 

The goal is to code an agent to solve this task that will learn by trial and error. We propose the following environment:

In [4]:
class Environment(object):
    def __init__(self, grid_size=10, max_time=500, temperature=0.1):
        grid_size = grid_size+4
        self.grid_size = grid_size
        self.max_time = max_time
        self.temperature = temperature

        #board on which one plays
        self.board = np.zeros((grid_size,grid_size))
        self.position = np.zeros((grid_size,grid_size))

        # coordinate of the cat
        self.x = 0
        self.y = 1

        # self time
        self.t = 0

        self.scale=16

        self.to_draw = np.zeros((max_time+2, grid_size*self.scale, grid_size*self.scale, 3))


    def draw(self,e):
        skvideo.io.vwrite(str(e) + '.mp4', self.to_draw)

    def get_frame(self,t):
        b = np.zeros((self.grid_size,self.grid_size,3))+128
        b[self.board>0,0] = 256
        b[self.board < 0, 2] = 256
        b[self.x,self.y,:]=256
        b[-2:,:,:]=0
        b[:,-2:,:]=0
        b[:2,:,:]=0
        b[:,:2,:]=0
        
        b =  cv2.resize(b, None, fx=self.scale, fy=self.scale, interpolation=cv2.INTER_NEAREST)

        self.to_draw[t,:,:,:]=b


    def act(self, action):
        """This function returns the new state, reward and decides if the
        game ends."""

        self.get_frame(int(self.t))

        self.position = np.zeros((self.grid_size, self.grid_size))

        self.position[0:2,:]= -1
        self.position[:,0:2] = -1
        self.position[-2:, :] = -1
        self.position[-2:, :] = -1

        self.position[self.x, self.y] = 1
        if action == 0:
            if self.x == self.grid_size-3:
                self.x = self.x-1
            else:
                self.x = self.x + 1
        elif action == 1:
            if self.x == 2:
                self.x = self.x+1
            else:
                self.x = self.x-1
        elif action == 2:
            if self.y == self.grid_size - 3:
                self.y = self.y - 1
            else:
                self.y = self.y + 1
        elif action == 3:
            if self.y == 2:
                self.y = self.y + 1
            else:
                self.y = self.y - 1
        else:
            RuntimeError('Error: action not recognized')

        self.t = self.t + 1
        reward = self.board[self.x, self.y]
        self.board[self.x, self.y] = 0
        game_over = self.t > self.max_time
        state = np.concatenate((self.board.reshape(self.grid_size, self.grid_size,1),
                        self.position.reshape(self.grid_size, self.grid_size,1)),axis=2)
        state = state[self.x-2:self.x+3,self.y-2:self.y+3,:]

        return state, reward, game_over

    def reset(self):
        """This function resets the game and returns the initial state"""

        self.x = np.random.randint(3, self.grid_size-3, size=1)[0]
        self.y = np.random.randint(3, self.grid_size-3, size=1)[0]


        bonus = 0.5*np.random.binomial(1,self.temperature,size=self.grid_size**2)
        bonus = bonus.reshape(self.grid_size,self.grid_size)

        malus = -1.0*np.random.binomial(1,self.temperature,size=self.grid_size**2)
        malus = malus.reshape(self.grid_size, self.grid_size)

        self.to_draw = np.zeros((self.max_time+2, self.grid_size*self.scale, self.grid_size*self.scale, 3))


        malus[bonus>0]=0

        self.board = bonus + malus

        self.position = np.zeros((self.grid_size, self.grid_size))
        self.position[0:2,:]= -1
        self.position[:,0:2] = -1
        self.position[-2:, :] = -1
        self.position[-2:, :] = -1
        self.board[self.x,self.y] = 0
        self.t = 0

        state = np.concatenate((
                               self.board.reshape(self.grid_size, self.grid_size,1),
                        self.position.reshape(self.grid_size, self.grid_size,1)),axis=2)

        state = state[self.x - 2:self.x + 3, self.y - 2:self.y + 3, :]
        return state

The following elements are important because they correspond to the hyper parameters for this project:

In [5]:
# parameters
size = 13
T=200
temperature=0.3
epochs_train=100 # set small when debugging
epochs_test=10 # set small when debugging

# display videos
def display_videos(name):
    video = io.open(name, 'r+b').read()
    encoded = base64.b64encode(video)
    return '''<video alt="test" controls>
                <source src="data:video/mp4;base64,{0}" type="video/mp4" />
             </video>'''.format(encoded.decode('ascii'))

__Question 2__ Explain the use of the arrays ```position``` and ```board```.

The array position represents the possibility for the rat movement :
- -1 where the rat can't go.
- 0 the available position where the rat can move
- 1 the current position of the rat

The array board represents the reward of each cell.

## Random Agent

***
__Question 3__ Implement a random Agent (only ```learned_act``` needs to be implemented):

In [6]:
class RandomAgent(Agent):
    def __init__(self):
        super(RandomAgent, self).__init__()
        pass

    def learned_act(self, s):
        return np.random.randint(0,self.n_action)
        pass

***
***
__Question 4__ Visualize the game moves. You need to fill in the following function for the evaluation:

In [7]:
def test(agent,env,epochs,prefix=''):
    # Number of won games
    score = 0
        
    for e in range(epochs):
        
        ##### FILL IN HERE
        state = env.reset()
        game_over = False

        win = 0
        lose = 0
        
        
        while not game_over:
            
            action = agent.learned_act(state)
            state,reward,game_over = env.act(action)
            if reward >0:
                win = win +reward 
            if reward < 0:
                lose = lose-reward
                
        # Save as a mp4
        env.draw(prefix+str(e))

        # Update stats
        score = score + win-lose

        print("Win/lose count {}/{}. Average score ({})"
              .format(win, lose, score/(1+e)))
    print('Final score: '+str(score/epochs))

In [8]:
# Initialize the game
env = Environment(grid_size=size, max_time=T,temperature=temperature)

# Initialize the agent!
agent = RandomAgent()

test(agent,env,epochs_test,prefix='random')
HTML(display_videos('random0.mp4'))

Win/lose count 7.0/9.0. Average score (-2.0)
Win/lose count 11.5/16.0. Average score (-3.25)
Win/lose count 11.0/17.0. Average score (-4.166666666666667)
Win/lose count 8.5/18.0. Average score (-5.5)
Win/lose count 14.0/7.0. Average score (-3.0)
Win/lose count 11.0/17.0. Average score (-3.5)
Win/lose count 10.0/18.0. Average score (-4.142857142857143)
Win/lose count 10.5/11.0. Average score (-3.6875)
Win/lose count 10.0/14.0. Average score (-3.7222222222222223)
Win/lose count 5.0/14.0. Average score (-4.25)
Final score: -4.25


***
## DQN

Let us assume here that $T=\infty$.

***
__Question 5__ Let $\pi$ be a policy, show that:

\begin{equation*}
Q^{\pi}(s,a)=E_{(s',a')\sim p(.|s,a)}[r(s,a)+\gamma Q^{\pi}(s',a')]
\end{equation*}

Then, show that for the optimal policy $\pi^*$ (we assume its existence), the following holds: 

\begin{equation*}
Q^{*}(s,a)=E_{s'\sim \pi^*(.|s,a)}[r(s,a)+\gamma\max_{a'}Q^{*}(s',a')].
\end{equation*}
Finally, deduce that a plausible objective is:

\begin{equation*}
\mathcal{L}(\theta)=E_{s' \sim \pi^*(.|s,a)}\Vert r+\gamma\max\max_{a'}Q(s',a',\theta)-Q(s,a,\theta)\Vert^{2}.
\end{equation*}




For a policy $\pi$, a state $s$ and an action $a$ :
$$\begin{align*}
Q^{\pi}(s,a) &= E_{p^{\pi}}\left[\sum_{t\leq T}\gamma^{t}r(s_{t},a_{t})|s_{0}=s,a_{0}=a\right]  \\
&= r(s,a) + \gamma \sum_{s'}E_{p^{\pi}}\left[\sum_{t= 1}^{T}\gamma^{t-1}r(s_{t},a_{t})\mathbb{1}_{\{s_{1} = s'\}}|s_{0}=s,a_{0}=a\right] \\
&=  r(s,a) + \gamma \sum_{s'}\mathbb{P}(s'|s,a)E_{p^{\pi}}\left[\sum_{t= 1}^{T}\gamma^{t-1}r(s_{t},a_{t}) | s_{1}=s'\right] \\
 & = r(s,a) + \gamma \sum_{s',a'}\mathbb{P}(s'|s,a)\pi(a'|s')E_{p^{\pi}}\left[\sum_{t= 1}^{T}\gamma^{t-1}r(s_{t},a_{t}) | s_{1}=s',a_{1}=a'\right] \\
&=  r(s,a) + \gamma \mathbb{E}_{(s',a')\sim (\mathbb{P}(.|s,a),\pi(.|s'))}(Q^{\pi}(s',a')) \\
\end{align*}$$

The optimal $Q$-function is: $Q^*(s,a)=\max_{\pi}Q^\pi(s,a)$. Thus, this is equivalent to maximizing the second term with respect to $a^\prime$. 

Then: $Q^*(s,a)= r(s,a) + \gamma E_{(s^\prime, a^\prime)\sim p^{\pi^*}(.|s,a)} [\max_{a^\prime}Q^*(s^\prime,a^\prime)]$.

***
The DQN-learning algorithm relies on these derivations to train the parameters $\theta$ of a Deep Neural Network:

1. At the state $s_t$, select the action $a_t$ with best reward using $Q_t$ and store the results;

2. Obtain the new state $s_{t+1}$ from the environment $p$;

3. Store $(s_t,a_t,s_{t+1})$;

4. Obtain $Q_{t+1}$ by minimizing  $\mathcal{L}$ from a recovered batch from the previously stored results.

***
__Question 6__ Implement the class ```Memory``` that stores moves (in a replay buffer) via ```remember``` and provides a ```random_access``` to these. Specify a maximum memory size to avoid side effects. You can for example use a ```list()``` and set by default ```max_memory=100```.

In [9]:
class Memory(object):
    def __init__(self, max_memory=100):
        self.max_memory = max_memory
        self.memory = list()

    def remember(self, m):
        self.memory.append(m)
        if len(self.memory)>= self.max_memory:
            del self.memory[0]

    def random_access(self):
        i = np.random.randint(0,len(self.memory))
        return self.memory[i]

***
The pipeline we will use for training is given below:

In [10]:
def train(agent,env,epoch,prefix=''):
    # Number of won games
    score = 0
    loss = 0

    for e in range(epoch):
        # At each epoch, we restart to a fresh game and get the initial state
        state = env.reset()
        # This assumes that the games will terminate
        game_over = False

        win = 0
        lose = 0

        while not game_over:
            # The agent performs an action
            action = agent.act(state)

            # Apply an action to the environment, get the next state, the reward
            # and if the games end
            prev_state = state
            state, reward, game_over = env.act(action)

            # Update the counters
            if reward > 0:
                win = win + reward
            if reward < 0:
                lose = lose -reward

            # Apply the reinforcement strategy
            loss = agent.reinforce(prev_state, state,  action, reward, game_over)

        # Save as a mp4
        if e % 10 == 0:
            env.draw(prefix+str(e))

        # Update stats
        score += win-lose

        print("Epoch {:03d}/{:03d} | Loss {:.4f} | Win/lose count {}/{} ({})"
              .format(e, epoch, loss, win, lose, win-lose))
        agent.save(name_weights=prefix+'model.h5',name_model=prefix+'model.json')

***
__Question 7__ Implement the DQN training algorithm using a cascade of fully connected layers. You can use different learning rate, batch size or memory size parameters. In particular, the loss might oscillate while the player will start to win the games. You have to find a good criterium.

In [14]:
class DQN(Agent):
    def __init__(self, grid_size,  epsilon = 0.1, memory_size=100, batch_size = 16,n_state=2):
        super(DQN, self).__init__(epsilon = epsilon)

        # Discount for Q learning
        self.discount = 0.99
        
        self.grid_size = grid_size
        
        # number of state
        self.n_state = n_state

        # Memory
        self.memory = Memory(memory_size)
        
        # Batch size when learning
        self.batch_size = batch_size

    def learned_act(self, s):
        pred = self.model.predict(s[None])
        return np.argmax(pred)


    def reinforce(self, s_, n_s_, a_, r_, game_over_):
        # Two steps: first memorize the states, second learn from the pool

        self.memory.remember([s_, n_s_, a_, r_, game_over_])
        
        input_states = np.zeros((self.batch_size, 5,5,self.n_state))
        target_q = np.zeros((self.batch_size, 4))
        
        for i in range(self.batch_size):
            s, n_s, a, r, game_over = self.memory.random_access()
            target_q[i] = self.model.predict(s[None,:])

            if game_over:
                target_q[i][a] = r
            else:
                target_q[i][a] = r + self.discount*max((self.model.predict(n_s[None,:]))[0])
            
            
            input_states[i,:] = s 

        # HINT: Clip the target to avoid exploiding gradients.. -- clipping is a bit tighter
        target_q = np.clip(target_q, -3, 3)

        l = self.model.train_on_batch(input_states, target_q)


        return l

    def save(self,name_weights='model.h5',name_model='model.json'):
        self.model.save_weights(name_weights, overwrite=True)
        with open(name_model, "w") as outfile:
            json.dump(self.model.to_json(), outfile)
            
    def load(self,name_weights='model.h5',name_model='model.json'):
        with open(name_model, "r") as jfile:
            model = model_from_json(json.load(jfile))
        model.load_weights(name_weights)
        model.compile("sgd", "mse")
        self.model = model

            
class DQN_FC(DQN):
    def __init__(self, *args, lr=0.1,**kwargs):
        super(DQN_FC, self).__init__( *args,**kwargs)
        
        # NN Model

        model = Sequential()
        model.add(Flatten(input_shape = (5,5,self.n_state)))
        model.add(Dense(32,activation  = 'relu'))
        model.add(Dense(4))
        
        model.compile(sgd(lr=lr, decay=1e-4, momentum=0.0), "mse")
        self.model = model
        

In [15]:
env = Environment(grid_size=size, max_time=T, temperature=0.3)
agent = DQN_FC(size, lr=.1, epsilon = 0.1, memory_size=2000, batch_size = 32)
train(agent, env, epochs_train, prefix='fc_train')
HTML(display_videos('fc_train10.mp4'))

Epoch 000/100 | Loss 0.0349 | Win/lose count 5.0/11.0 (-6.0)
Epoch 001/100 | Loss 0.0165 | Win/lose count 1.5/7.0 (-5.5)
Epoch 002/100 | Loss 0.0106 | Win/lose count 2.5/1.0 (1.5)
Epoch 003/100 | Loss 0.0084 | Win/lose count 5.5/6.0 (-0.5)
Epoch 004/100 | Loss 0.0064 | Win/lose count 7.5/4.0 (3.5)
Epoch 005/100 | Loss 0.0147 | Win/lose count 7.5/8.0 (-0.5)
Epoch 006/100 | Loss 0.0379 | Win/lose count 6.0/3.0 (3.0)
Epoch 007/100 | Loss 0.0254 | Win/lose count 2.5/6.0 (-3.5)
Epoch 008/100 | Loss 0.0875 | Win/lose count 10.0/6.0 (4.0)
Epoch 009/100 | Loss 0.0087 | Win/lose count 2.0/3.0 (-1.0)
Epoch 010/100 | Loss 0.0109 | Win/lose count 9.5/2.0 (7.5)
Epoch 011/100 | Loss 0.0047 | Win/lose count 3.0/5.0 (-2.0)
Epoch 012/100 | Loss 0.0124 | Win/lose count 1.5/1.0 (0.5)
Epoch 013/100 | Loss 0.0083 | Win/lose count 3.0/2.0 (1.0)
Epoch 014/100 | Loss 0.0047 | Win/lose count 3.0/5.0 (-2.0)
Epoch 015/100 | Loss 0.0038 | Win/lose count 3.5/5.0 (-1.5)
Epoch 016/100 | Loss 0.0551 | Win/lose count 

***
***
__Question 8__ Implement the DQN training algorithm using a CNN (for example, 2 convolutional layers and one final fully connected layer).

In [16]:
class DQN_CNN(DQN):
    def __init__(self, *args,lr=0.1,**kwargs):
        super(DQN_CNN, self).__init__(*args,**kwargs)
        
        model = Sequential()
        model.add(Conv2D(32,kernel_size = (2,2), activation = 'relu',input_shape = (5,5,self.n_state)))
        model.add(Conv2D(32,kernel_size = (2,2), activation = 'relu'))
        model.add(Flatten())
        model.add(Dense(4))
        
        model.compile(sgd(lr=lr, decay=1e-4, momentum=0.0), "mse")
        self.model = model

In [17]:
env = Environment(grid_size=size, max_time=T, temperature=0.3)
agent = DQN_CNN(size, lr=.1, epsilon = 0.1, memory_size=2000, batch_size = 32)
train(agent,env,epochs_train,prefix='cnn_train')
HTML(display_videos('cnn_train10.mp4'))

Epoch 000/100 | Loss 0.0037 | Win/lose count 1.0/1.0 (0.0)
Epoch 001/100 | Loss 0.0039 | Win/lose count 3.5/1.0 (2.5)
Epoch 002/100 | Loss 0.0311 | Win/lose count 4.5/3.0 (1.5)
Epoch 003/100 | Loss 0.0048 | Win/lose count 4.5/4.0 (0.5)
Epoch 004/100 | Loss 0.0061 | Win/lose count 4.5/3.0 (1.5)
Epoch 005/100 | Loss 0.0010 | Win/lose count 1.5/5.0 (-3.5)
Epoch 006/100 | Loss 0.0020 | Win/lose count 6.0/3.0 (3.0)
Epoch 007/100 | Loss 0.0058 | Win/lose count 8.5/3.0 (5.5)
Epoch 008/100 | Loss 0.0122 | Win/lose count 10.0/1.0 (9.0)
Epoch 009/100 | Loss 0.0042 | Win/lose count 9.0/8.0 (1.0)
Epoch 010/100 | Loss 0.0452 | Win/lose count 9.5/3.0 (6.5)
Epoch 011/100 | Loss 0.0048 | Win/lose count 7.5/5.0 (2.5)
Epoch 012/100 | Loss 0.0047 | Win/lose count 10.5/2.0 (8.5)
Epoch 013/100 | Loss 0.0064 | Win/lose count 3.5/1.0 (2.5)
Epoch 014/100 | Loss 0.0043 | Win/lose count 4.0/2.0 (2.0)
Epoch 015/100 | Loss 0.0022 | Win/lose count 8.5/2.0 (6.5)
Epoch 016/100 | Loss 0.0487 | Win/lose count 12.0/1.0

***
***
__Question 9__ Test both algorithms and compare their performances. Which issue(s) do you observe? Observe also different behaviors by changing the temperature.

In [21]:
env = Environment(grid_size=size, max_time=T,temperature=0.3)
agent_cnn = DQN_CNN(size, lr=.1, epsilon = 0.1, memory_size=2000, batch_size = 32)
agent_cnn.load(name_weights='cnn_trainmodel.h5',name_model='cnn_trainmodel.json')

agent_fc = DQN_FC(size, lr=.1, epsilon = 0.1, memory_size=2000, batch_size = 32)
agent_cnn.load(name_weights='fc_trainmodel.h5',name_model='fc_trainmodel.json')
print('Test of the CNN')
test(agent_cnn,env,epochs_test,prefix='cnn_test')
print('Test of the FC')
test(agent_fc,env,epochs_test,prefix='fc_test')

Test of the CNN
Win/lose count 3.0/0. Average score (3.0)
Win/lose count 7.0/0. Average score (5.0)
Win/lose count 10.5/0. Average score (6.833333333333333)
Win/lose count 1.0/0. Average score (5.375)
Win/lose count 8.0/0. Average score (5.9)
Win/lose count 17.0/0. Average score (7.75)
Win/lose count 5.0/0. Average score (7.357142857142857)
Win/lose count 3.5/0. Average score (6.875)
Win/lose count 5.0/0. Average score (6.666666666666667)
Win/lose count 7.0/0. Average score (6.7)
Final score: 6.7
Test of the FC
Win/lose count 0/0. Average score (0.0)
Win/lose count 0/3.0. Average score (-1.5)
Win/lose count 0/1.0. Average score (-1.3333333333333333)
Win/lose count 4.0/3.0. Average score (-0.75)
Win/lose count 4.0/7.0. Average score (-1.2)
Win/lose count 0.5/2.0. Average score (-1.25)
Win/lose count 1.5/1.0. Average score (-1.0)
Win/lose count 0.5/2.0. Average score (-1.0625)
Win/lose count 2.0/3.0. Average score (-1.0555555555555556)
Win/lose count 0.5/0. Average score (-0.9)
Final sco

In [31]:
HTML(display_videos('cnn_test9.mp4'))

In [32]:
HTML(display_videos('fc_test9.mp4'))

We can conclude that the CNN architecture gives better result that the one using connected networks because of overfiting.
With a higher temperature the gap is higher.

Broadly speaking, we can notice that the main issue is that the algorithm is stucked on a local minima and therefore can't explore the entire map.

***

The algorithm tends to not explore the map which can be an issue. We propose two ideas in order to encourage exploration:
1. Incorporating a decreasing $\epsilon$-greedy exploration. You can use the method ```set_epsilon```
2. Append via the environment a new state that describes if a cell has been visited or not

***
__Question 10__ Design a new ```train_explore``` function and environment class ```EnvironmentExploring``` to tackle the issue of exploration.



In [27]:
def train_explore(agent,env,epoch,prefix=''):
    score = 0
    loss = 0

    for e in range(epoch):
      
        state = env.reset()
        game_over = False

        win = 0
        lose = 0
        agent.set_epsilon(agent.epsilon*0.90)
        while not game_over:
         
            action = agent.act(state)

            
            prev_state = state
            state, reward, game_over = env.act(action, train=True)

            
            if reward > 0:
                win = win + reward
            if reward < 0:
                lose = lose -reward

            
            loss = agent.reinforce(prev_state, state,  action, reward, game_over)
            
      
        if e % 10 == 0:
              env.draw(prefix+str(e))

       
        score += win-lose

        print("Epoch {:03d}/{:03d} | Loss {:.4f} | Win/lose count {}/{} ({})"
              .format(e, epoch, loss, win, lose, win-lose))
        agent.save(name_weights=prefix+'model.h5',name_model=prefix+'model.json')
        
class EnvironmentExploring(object):
    def __init__(self, grid_size=10, max_time=500, temperature=0.1):
        grid_size = grid_size+4
        self.grid_size = grid_size
        self.max_time = max_time
        self.temperature = temperature

        
        self.board = np.zeros((grid_size,grid_size))
        self.position = np.zeros((grid_size,grid_size))
        self.malus_position = np.zeros((grid_size,grid_size)) 
        self.x = 0
        self.y = 1

        
        self.t = 0

        self.scale=16

        self.to_draw = np.zeros((max_time+2, grid_size*self.scale, grid_size*self.scale, 3))


    def draw(self,e):
        skvideo.io.vwrite(str(e) + '.mp4', self.to_draw)
        
        
    def get_frame(self,t):
        b = np.zeros((self.grid_size,self.grid_size,3))+128
        b[self.board>0,0] = 256
        b[self.board < 0, 2] = 256
        b[self.x,self.y,:]=256
        b[-2:,:,:]=0
        b[:,-2:,:]=0
        b[:2,:,:]=0
        b[:,:2,:]=0
        
        b =  cv2.resize(b, None, fx=self.scale, fy=self.scale, interpolation=cv2.INTER_NEAREST)

        self.to_draw[t,:,:,:]=b
    def act(self, action,train = False):
        

        self.get_frame(int(self.t))

        self.position = np.zeros((self.grid_size, self.grid_size))

        self.position[0:2,:]= -1
        self.position[:,0:2] = -1
        self.position[-2:, :] = -1
        self.position[:,-2:] = -1

        self.position[self.x, self.y] = 1
        if action == 0:
            if self.x == self.grid_size-3:
                self.x = self.x-1
            else:
                self.x = self.x + 1
        elif action == 1:
            if self.x == 2:
                self.x = self.x+1
            else:
                self.x = self.x-1
        elif action == 2:
            if self.y == self.grid_size - 3:
                self.y = self.y - 1
            else:
                self.y = self.y + 1
        elif action == 3:
            if self.y == 2:
                self.y = self.y + 1
            else:
                self.y = self.y - 1
        else:
            RuntimeError('Error: action not recognized')

        self.t = self.t + 1
        reward = 0
        if train:
            reward = -self.malus_position[self.x, self.y]
        self.malus_position[self.x, self.y] = 0.1
        reward = reward + self.board[self.x, self.y]
        self.board[self.x, self.y] = 0
        game_over = self.t > self.max_time
        
        
        state = np.concatenate((self.malus_position.reshape(self.grid_size, self.grid_size,1),
                                self.board.reshape(self.grid_size, self.grid_size,1),
                        self.position.reshape(self.grid_size, self.grid_size,1)),axis=2)
        state = state[self.x-2:self.x+3,self.y-2:self.y+3,:]

        return state, reward, game_over
      
    def reset(self):
        """This function resets the game and returns the initial state"""

        self.x = np.random.randint(3, self.grid_size-3, size=1)[0]
        self.y = np.random.randint(3, self.grid_size-3, size=1)[0]


        bonus = 0.5*np.random.binomial(1,self.temperature,size=self.grid_size**2)
        bonus = bonus.reshape(self.grid_size,self.grid_size)

        malus = -1.0*np.random.binomial(1,self.temperature,size=self.grid_size**2)
        malus = malus.reshape(self.grid_size, self.grid_size)

        self.to_draw = np.zeros((self.max_time+2, self.grid_size*self.scale, self.grid_size*self.scale, 3))


        malus[bonus>0]=0

        self.board = bonus + malus
        self.malus_position = 0*self.board

        self.position = np.zeros((self.grid_size, self.grid_size))
        self.position[0:2,:]= -1
        self.position[:,0:2] = -1
        self.position[-2:, :] = -1
        self.position[:, -2:] = -1
        self.board[self.x,self.y] = 0
        self.t = 0
        # Modify the state variable to include the malus map
        state = np.concatenate((self.malus_position.reshape(self.grid_size, self.grid_size,1),
                                self.board.reshape(self.grid_size, self.grid_size,1),
                        self.position.reshape(self.grid_size, self.grid_size,1)),axis=2)
       
        
        state = state[self.x - 2:self.x + 3, self.y - 2:self.y + 3, :]
        return state

In [28]:
# Training
env = EnvironmentExploring(grid_size=size, max_time=T, temperature=0.3)
agent = DQN_CNN(size, lr=.1, epsilon = 0.1, memory_size=2000, batch_size = 32,n_state=3)
train_explore(agent, env, epochs_train, prefix='cnn_train_explore')
HTML(display_videos('cnn_train_explore10.mp4'))

Epoch 000/100 | Loss 0.0014 | Win/lose count 3.0/22.30000000000005 (-19.30000000000005)
Epoch 001/100 | Loss 0.0063 | Win/lose count 6.0/22.200000000000045 (-16.200000000000045)
Epoch 002/100 | Loss 0.0034 | Win/lose count 6.0/21.80000000000005 (-15.80000000000005)
Epoch 003/100 | Loss 0.0091 | Win/lose count 6.0/21.20000000000004 (-15.200000000000038)
Epoch 004/100 | Loss 0.0056 | Win/lose count 9.5/20.00000000000003 (-10.500000000000028)
Epoch 005/100 | Loss 0.0120 | Win/lose count 9.5/17.299999999999976 (-7.799999999999976)
Epoch 006/100 | Loss 0.0076 | Win/lose count 9.5/16.79999999999997 (-7.299999999999969)
Epoch 007/100 | Loss 0.0042 | Win/lose count 11.5/16.799999999999976 (-5.299999999999976)
Epoch 008/100 | Loss 0.0082 | Win/lose count 8.0/16.79999999999997 (-8.799999999999969)
Epoch 009/100 | Loss 0.0044 | Win/lose count 21.5/13.599999999999971 (7.900000000000029)
Epoch 010/100 | Loss 0.0088 | Win/lose count 15.5/16.799999999999972 (-1.2999999999999723)
Epoch 011/100 | Loss 

Epoch 092/100 | Loss 0.0165 | Win/lose count 25.0/12.899999999999977 (12.100000000000023)
Epoch 093/100 | Loss 0.0108 | Win/lose count 21.0/14.999999999999972 (6.000000000000028)
Epoch 094/100 | Loss 0.0175 | Win/lose count 27.0/11.999999999999973 (15.000000000000027)
Epoch 095/100 | Loss 0.0088 | Win/lose count 27.5/8.999999999999984 (18.500000000000014)
Epoch 096/100 | Loss 0.0226 | Win/lose count 18.0/12.299999999999978 (5.700000000000022)
Epoch 097/100 | Loss 0.0093 | Win/lose count 23.5/12.699999999999976 (10.800000000000024)
Epoch 098/100 | Loss 0.0155 | Win/lose count 21.5/14.799999999999972 (6.700000000000028)
Epoch 099/100 | Loss 0.0204 | Win/lose count 25.5/10.299999999999985 (15.200000000000015)


In [30]:
# Evaluation
test(agent,env,epochs_test,prefix='cnn_test_explore')
HTML(display_videos('cnn_test_explore9.mp4'))

Win/lose count 28.5/6.0. Average score (22.5)
Win/lose count 21.0/2.0. Average score (20.75)
Win/lose count 24.5/8.0. Average score (19.333333333333332)
Win/lose count 24.5/5.0. Average score (19.375)
Win/lose count 28.5/4.0. Average score (20.4)
Win/lose count 14.5/2.0. Average score (19.083333333333332)
Win/lose count 20.5/3.0. Average score (18.857142857142858)
Win/lose count 22.0/3.0. Average score (18.875)
Win/lose count 22.5/2.0. Average score (19.055555555555557)
Win/lose count 24.0/4.0. Average score (19.15)
Final score: 19.15


We realy improved the strategy since we obtained a final score of 19.15

***
***
__BONUS question__ Use the expert DQN from the previous question to generate some winning games. Train a model that mimicks its behavior. Compare the performances.

***